doc/poddoc/man/gspool_open.3
.\" Automatically generated by Pod::Man 2.22 (Pod::Simple 3.13)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "gspool_open 3"
.TH gspool_open 3 "2009-05-30" "GNUspool Release 1" "GNUspool Print Manager"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
gspool_open \- open a connection to a GNUspool Host
.PP
gspool_login \- log in to GNUspool Host from Windows
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fB#include <gspool.h>\fR
.PP
\&\fBint gspool_open(const char *\fR\fIhost\fR\fB, const char *\fR\fIserv\fR\fB, const classcode_t\fR \fIcl\fR\fB)\fR
<br>
\&\fBint gspool_open(const char *\fR\fIhost\fR\fB, const char *\fR\fIserv\fR\fB, const char *\fR\fIuser\fR\fB, const classcode_t\fR \fIcl\fR\fB)\fR
<br>
\&\fBint gspool_login(const char *\fR\fIhost\fR\fB, const char *\fR\fIserv\fR\fB, const char *\fR\fIuser\fR\fB, char *\fR\fIpass\fR\fB, const classcode_t\fR \fIcl\fR\fB)\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
The \fBgspool_open\fR routine is used to commence a session with the
\&\s-1API\s0. The first form is that used on Unix or GNU/Linux hosts, the user
name being obtained from the effective user id invoking the routine.
.PP
The second form is that used on Windows hosts, where the Windows
machine has a static \s-1IP\s0 address and no password is required.
.PP
The third form is that used on Windows hosts with \s-1IP\s0 addresses
allocated via \s-1DHCP\s0 and/or a password is required.
.PP
\&\fIhost\fR is the name of the host to be connected to. \fIServ\fR is the
name of the \s-1TCP/UDP\s0 service to be used to connect to \fBGNUspool\fR.  If
this is set to \s-1NULL\s0, the default service will be used (but this must
be set up correctly).
.PP
\&\fIuser\fR is the Unix user name to be used on Windows hosts and \fIpass\fR
is the password. Note that this is not \f(CW\*(C`const\*(C'\fR, and the passed field
will be deliberately overwritten as soon as possible within
\&\fBgspool_login\fR. Also note that the password may not be the Unix password,
it may be an \fIinterface password\fR set up by \fIxipasswd\fR\|(8).
.PP
\&\fIcl\fR is the classcode to be used for access to all printer and job
operations. If 0 is used the user's default class code will be
used. Note that the class code bits are assigned starting at the least
significant bit, thus:
.PP
.Vb 10
\& 1 = A
\& 2 = B
\& 4 = C
\& ....
\& 0x8000 = P
\& 0x10000 = a
\& 0x20000 = b
\& 0x40000 = c
\& ....
\& 0x80000000 = p
.Ve
.SH "RETURN VALUES"
.IX Header "RETURN VALUES"
The functions return an integer descriptor on success, which will be
greater than or equal to 0. This descriptor should be used in all
subsequent operations with the \s-1API\s0.
If an error occurs, one of the following (negative) error codes will be returned.
.IP "GSPOOL_BADWRITE" 8
failure writing to the network
.IP "GSPOOL_BADREAD" 8
failure reading from the network
.IP "GSPOOL_NOMEM" 8
out of memory accessing network
.IP "GSPOOL_UNKNOWN_USER" 8
user name not known
.IP "GSPOOL_NO_PASSWD" 8
password must be specified
.IP "GSPOOL_PASSWD_INVALID" 8
invalid password
.PP
Each connection should also be terminated by a call to \fBgspool_close\fR or
exit from the calling program.
.SH "NOTES"
.IX Header "NOTES"
Any number of \s-1API\s0 sesssions, to the same host or to various hosts, may
be in progress simultaneously, subject to the restrictions on the
number of simultaneous open files per process which the operating
system allows.
.SH "EXAMPLE"
.IX Header "EXAMPLE"
An example to open a connection to host \*(L"xisl\*(R":
.PP
.Vb 5
\& int fd, ret;
\& fd = gspool_open("xisl", (char *) 0, 0);
\& if (fd < 0) { /* error handling */
\&     ...
\& }
\&
\& /* process connection */
\&
\& gspool_close(fd);
.Ve
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIgspool_close\fR\|(3),
\&\fIgspool_jobmon\fR\|(3),
\&\fIgspool_ptrmon\fR\|(3).
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Copyright (c) 2009 Free Software Foundation, Inc.
This is free software. You may redistribute copies of it under the
terms of the \s-1GNU\s0 General Public License
<http://www.gnu.org/licenses/gpl.html>.
There is \s-1NO\s0 \s-1WARRANTY\s0, to the extent permitted by law.
.SH "AUTHOR"
.IX Header "AUTHOR"
John M Collins, Xi Software Ltd.


debian/changelog
gnuspool (1.7) unstable; urgency=low

  * strip out unneeded sections of postinst etc. and do some related tidying
  * remove MSWIN subdirectory and bump version in configure

 -- John M Collins (Xi Software Ltd) <jmc@xisl.com>  Mon, 26 Apr 2010 12:26:21 +0000

gnuspool (1.6) unstable; urgency=low

  [ John M Collins (Xi Software Ltd) ]
  * Initial release (Closes: #556996)

 -- John M Collins (Xi Software Ltd) <jmc@xisl.com>  Tue, 17 Nov 2009 10:16:21 +0000


src/apilib/gspool_open.c
/* gspool_open.c -- API open connection

   Copyright 2008 Free Software Foundation, Inc.

   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.  */

#include <stdio.h>
#include <sys/types.h>
#include "gspool.h"
#ifdef	HAVE_FCNTL_H
#include <fcntl.h>
#endif
#include <pwd.h>
#include <errno.h>
#ifdef	HAVE_SYS_IOCTL_H
#include <sys/ioctl.h>
#endif
#include "xtapi_int.h"
#include "incl_unix.h"
#include "incl_net.h"
#include "incl_sig.h"
#include "files.h"
#ifdef	TIME_WITH_SYS_TIME
#include <sys/time.h>
#include <time.h>
#elif	defined(HAVE_SYS_TIME_H)
#include <sys/time.h>
#else
#include <time.h>
#endif
#include "services.h"

#ifndef	_NFILE
#define	_NFILE	64
#endif

#define	MAXFDS	(_NFILE / 3)

static	unsigned	api_max;
static	struct	api_fd	apilist[MAXFDS];

int	gspool_dataerror;

int	gspool_write(const int fd, char *buff, unsigned size)
{
	int	obytes;
	while  (size != 0)  {
		if  ((obytes = write(fd, buff, size)) < 0)  {
			if  (errno == EINTR)
				continue;
			return  GSPOOL_BADWRITE;
		}
		size -= obytes;
		buff += obytes;
	}
	return  0;
}

int	gspool_read(const int fd, char *buff, unsigned size)
{
	int	ibytes;
	while  (size != 0)  {
		if  ((ibytes = read(fd, buff, size)) < 0)  {
			if  (errno == EINTR)
				continue;
			return  GSPOOL_BADREAD;
		}
		size -= ibytes;
		buff += ibytes;
	}
	return  0;
}

int	gspool_wmsg(const struct api_fd *fdp, struct api_msg *msg)
{
	return  gspool_write(fdp->sockfd, (char *) msg, sizeof(struct api_msg));
}

int	gspool_rmsg(const struct api_fd *fdp, struct api_msg *msg)
{
	return  gspool_read(fdp->sockfd, (char *) msg, sizeof(struct api_msg));
}

struct	api_fd  *gspool_look_fd(const int fd)
{
	struct	api_fd	*result;

	if  (fd < 0  ||  (unsigned) fd >= api_max)
		return  (struct api_fd *) 0;
	result = &apilist[fd];
	if  (result->sockfd < 0)
		return  (struct api_fd *) 0;
	return  result;
}

#if defined(FASYNC) && defined(F_SETOWN)
static	struct api_fd *find_prod(const int fd)
{
	unsigned  cnt;
	for  (cnt = 0;  cnt < api_max;  cnt++)
		if  (apilist[cnt].prodfd == fd)
			return  &apilist[cnt];
	return  (struct api_fd *) 0;
}
#endif

int	gspool_open(const char *hostname, const char *servname, const classcode_t classcode)
{
	int	portnum, sock, ret;
	unsigned	result;
	struct	api_fd	*ret_fd;
	netid_t	hostid;
	const	char	*serv;
	struct	passwd	*pw;
	struct	hostent	*hp;
	struct	servent	*sp;
	struct	sockaddr_in	sin;
	struct	api_msg	outmsg;

	hp = gethostbyname(hostname);
	if  (!hp)  {
		endhostent();
		return  GSPOOL_INVALID_HOSTNAME;
	}
	hostid = * (LONG *) hp->h_addr;
	endhostent();

	serv = servname? servname: API_DEFAULT_SERVICE;
	if  (!(sp = getservbyname(serv, "tcp")))
		sp = getservbyname(serv, "TCP");
	if  (!sp)  {
		endservent();
		return  servname? GSPOOL_INVALID_SERVICE: GSPOOL_NODEFAULT_SERVICE;
	}
	portnum = ntohs(sp->s_port);
	endservent();

	/* And now for the groovy socket stuff....  */

	sin.sin_family = AF_INET;
	sin.sin_port = htons(portnum);
	BLOCK_ZERO(sin.sin_zero, sizeof(sin.sin_zero));
	sin.sin_addr.s_addr = hostid;

	if  ((sock = socket(PF_INET, SOCK_STREAM, 0)) < 0)
		return  GSPOOL_NOSOCKET;

	if  (connect(sock, (struct sockaddr *) &sin, sizeof(sin)) < 0)  {
		close(sock);
		return  GSPOOL_NOCONNECT;
	}


	/* Allocate ourselves a "file descriptor" and stuff our stuff in it.  */

	for  (result = 0;  result < api_max;  result++)
		if  (apilist[result].sockfd < 0)
			goto  found;
	if  (++api_max >= MAXFDS)
		return  GSPOOL_NOMEM;
 found:
	ret_fd = &apilist[result];
	ret_fd->portnum = (SHORT) portnum;
	ret_fd->sockfd = (SHORT) sock;
	ret_fd->prodfd = -1;
	ret_fd->classcode = classcode;
	ret_fd->hostid = hostid;
	ret_fd->jobfn = (void (*)()) 0;
	ret_fd->ptrfn = (void (*)()) 0;
	ret_fd->jserial = 0;
	ret_fd->pserial = 0;
	ret_fd->bufmax = 0;
	ret_fd->buff = (char *) 0;
	pw = getpwuid(geteuid());
	strncpy(ret_fd->username, pw? pw->pw_name: SPUNAME, UIDSIZE);
	ret_fd->username[UIDSIZE] = '\0';
	outmsg.code = API_SIGNON;
	strcpy(outmsg.un.signon.username, ret_fd->username);
	outmsg.un.signon.classcode = htonl(classcode);
	if  ((ret = gspool_wmsg(ret_fd, &outmsg)))  {
		gspool_close((int) result);
		return  ret;
	}
	if  ((ret = gspool_rmsg(ret_fd, &outmsg)))  {
		gspool_close((int) result);
		return  ret;
	}
	if  (outmsg.retcode != 0)  {
		gspool_close((int) result);
		return  (SHORT) ntohs(outmsg.retcode);
	}
	ret_fd->classcode = ntohl(outmsg.un.r_signon.classcode);
	return  (int) result;
}

#if	defined(FASYNC) && defined(F_SETOWN)
static	void	procpoll(int fd)
{
	struct	api_fd	*fdp;
	struct	api_msg	imsg;
	SOCKLEN_T		repl = sizeof(struct sockaddr_in);
	struct	sockaddr_in	reply_addr;

	if  (!(fdp = find_prod(fd)))
		return;

	if  (recvfrom(fdp->prodfd, (char *) &imsg, sizeof(imsg), 0, (struct sockaddr *) &reply_addr, &repl) < 0)
		return;
	switch  (imsg.code)  {
	default:
		return;
	case  API_JOBPROD:
		if  (fdp->jobfn)
			(*fdp->jobfn)((int) (fdp - apilist));
		fdp->jserial = ntohl(imsg.un.r_reader.seq);
		return;
	case  API_PTRPROD:
		if  (fdp->ptrfn)
			(*fdp->ptrfn)((int) (fdp - apilist));
		fdp->pserial = ntohl(imsg.un.r_reader.seq);
		return;
	}
}

static	RETSIGTYPE	catchpoll(int n)
{
	unsigned	cnt;
	int		fd;

	int	highfd = -1, nret;
	fd_set	ready;

#ifdef	UNSAFE_SIGNALS
	signal(n, catchpoll);
#endif

	for  (;;)  {
		FD_ZERO(&ready);
		for  (cnt = 0;  cnt < api_max;  cnt++)  {
			if  ((fd = apilist[cnt].prodfd) >= 0)  {
				FD_SET(fd, &ready);
				if  (fd > highfd)
					highfd = fd;
			}
		}
		nret = select(highfd+1, &ready, (fd_set *) 0, (fd_set *) 0, (struct timeval *) 0);
		if  (nret <= 0)
			break;
		for  (cnt = 0;  nret > 0  &&  cnt <= (unsigned) highfd;  cnt++)
			if  (FD_ISSET(cnt, &ready))  {
				procpoll(cnt);
				nret--;
			}
	}
}

static	int	setmon(struct api_fd *fdp)
{
	int	sockfd, pportnum, ret;
	const	char	*serv = API_MON_SERVICE;
	struct	servent	*sp;
#ifdef	STRUCT_SIG
	struct	sigstruct_name	z;
#endif
	int	protonum;
	struct	protoent	*pp;
	struct	sockaddr_in	sin;
	struct	api_msg	msg;

	if  (!(pp = getprotobyname("udp")) && !(pp = getprotobyname("UDP")))  {
		endprotoent();
		return  GSPOOL_NODEFAULT_SERVICE;
	}
	protonum = pp->p_proto;
	endprotoent();
	if  (!(sp = getservbyname(serv, "udp")))
		sp = getservbyname(serv, "UDP");
	if  (!sp)  {
		endservent();
		return  GSPOOL_NODEFAULT_SERVICE;
	}
	pportnum = sp->s_port;
	endservent();
	msg.code = API_REQPROD;
	if  ((ret = gspool_wmsg(fdp, &msg)))
		return  ret;
	sin.sin_family = AF_INET;
	sin.sin_port = (SHORT) pportnum;
	BLOCK_ZERO(sin.sin_zero, sizeof(sin.sin_zero));
	sin.sin_addr.s_addr = INADDR_ANY;

#ifdef	STRUCT_SIG
	z.sighandler_el = catchpoll;
	sigmask_clear(z);
	z.sigflags_el = SIGVEC_INTFLAG;
	sigact_routine(SIGIO, &z, (struct sigstruct_name *) 0);
#else
	signal(SIGIO, catchpoll);
#endif
	if  ((sockfd = socket(PF_INET, SOCK_DGRAM, protonum)) < 0)
		return  GSPOOL_NOSOCKET;
	if  (bind(sockfd, (struct sockaddr *) &sin, sizeof(sin)) < 0)  {
		close(sockfd);
		return  GSPOOL_NOBIND;
	}
	fdp->prodfd = sockfd;
	if  (fcntl(sockfd, F_SETOWN, getpid()) < 0  ||
	     fcntl(sockfd, F_SETFL, FASYNC) < 0)
		return  GSPOOL_NOSOCKET;
	return  GSPOOL_OK;
}

static	void	unsetmon(struct api_fd *fdp)
{
	struct	api_msg	msg;
	if  (fdp->prodfd < 0)
		return;
	close(fdp->prodfd);
	fdp->prodfd = -1;
	msg.code = API_UNREQPROD;
	gspool_wmsg(fdp, &msg);
}

int	gspool_jobmon(const int fd, void (*fn)(const int))
{
	struct	api_fd	*fdp = gspool_look_fd(fd);

	if  (!fdp)
		return  GSPOOL_INVALID_FD;

	if  (fn)  {
		int	ret;
		if  (fdp->prodfd < 0  &&  (ret = setmon(fdp)) != 0)
			return  ret;
	}
	else  if  (fdp->jobfn && !fdp->ptrfn)
		unsetmon(fdp);
	fdp->jobfn = fn;
	return  GSPOOL_OK;
}

int	gspool_ptrmon(const int fd, void (*fn)(const int))
{
	struct	api_fd	*fdp = gspool_look_fd(fd);

	if  (!fdp)
		return  GSPOOL_INVALID_FD;

	if  (fn)  {
		int	ret;
		if  (fdp->prodfd < 0  &&  (ret = setmon(fdp)) != 0)
			return  ret;
	}
	else  if  (fdp->ptrfn && !fdp->jobfn)
		unsetmon(fdp);
	fdp->ptrfn = fn;
	return  GSPOOL_OK;
}
#endif /* Possible to get SIGIO working */

int	gspool_close(const int fd)
{
	struct	api_fd	*fdp = gspool_look_fd(fd);
	struct	api_msg	outmsg;

	if  (!fdp)
		return  GSPOOL_INVALID_FD;
#if	defined(FASYNC) && defined(F_SETOWN)
	if  (fdp->prodfd >= 0)  {
		unsetmon(fdp);
		fdp->jobfn = (void (*)()) 0;
		fdp->ptrfn = (void (*)()) 0;
	}
#endif
	outmsg.code = API_SIGNOFF;
	gspool_wmsg(fdp, &outmsg);
	close(fdp->sockfd);
	fdp->sockfd = -1;
	if  (fdp->bufmax != 0)  {
		fdp->bufmax = 0;
		free(fdp->buff);
		fdp->buff = (char *) 0;
	}
	return  GSPOOL_OK;
}
